### PR TITLE
Moving high level graph pack/unpack Dask 

### PR DESCRIPTION
**Pull request by madsbk**
### Important: remove [CI hack](https://github.com/dask/dask/pull/7179/commits/c9bef90b29ff23a6d2c3a3df8b6be5036199bea0) before merge!

This PR moves all pack and unpacking of high level graphs from Distributed to Dask. This work is based on the discussion in https://github.com/dask/distributed/pull/4406. 
Beside the move, `unpack` now doesn't modify the input graph and dependencies, instead the unpacked graph and dependencies a returned, which makes it possible to delegate the pack and unpack of annotations to the base class. 

Notice, this PR includes a hack that makes CI install https://github.com/dask/distributed/pull/4489 before testing. This should be removed before merging.

- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`
- [x] Remove CI hack

cc. @sjperkins, @ian-r-rose, @jrbourbeau

### CODE REVIEW

File: dask/highlevelgraph.py
Comment (ian-r-rose) [inline dask/highlevelgraph.py:882]: These could be typed `MutableMapping`

**Review by ian-r-rose — COMMENTED** [review]
Thanks for taking this on @madsbk, this looks good to me

File: dask/highlevelgraph.py
Comment (madsbk) [reply dask/highlevelgraph.py]: Right, changed to `MutableMapping` :)

File: dask/highlevelgraph.py
Comment (crusaderky) [inline dask/highlevelgraph.py:384]: all_hlg_keys is of type Iterable, which doesn't have a .union method. AbstractSet in the function parameters + a change in variable name here is necessary.

```suggestion
        merged_hlg_keys = all_hlg_keys | dsk.keys()
        dsk = {
            stringify(k): stringify(v, exclusive=merged_hlg_keys) for k, v in dsk.items()
        }
```

File: dask/highlevelgraph.py
Comment (crusaderky) [inline dask/highlevelgraph.py:549]: nope, it was correct before (aligned to typeshed guidelines)

File: dask/highlevelgraph.py
Comment (crusaderky) [inline dask/highlevelgraph.py:883]:
```suggestion
    def __dask_distributed_pack__(self, client, client_keys: Iterable[Hashable]) -> Any:
```

File: dask/highlevelgraph.py
Comment (crusaderky) [inline dask/highlevelgraph.py:928]: No reason to be overly generic in return types. The only exception is when you want to return an internal state and make it read-only.
```suggestion
    ) -> Tuple[Dict[str, Any], Dict[str, set], Dict[str, Any]]:
```

File: dask/highlevelgraph.py
Comment (crusaderky) [inline dask/highlevelgraph.py:951]:
```suggestion
        dsk: Dict[str, Any]
            Materialized (stringified) graph of all nodes in the high level graph
        deps: Dict[str, set]
            Dependencies of each key in `dsk`
        anno: Dict[str, Any]
            Annotations for `dsk`
```

File: dask/highlevelgraph.py
Comment (crusaderky) [inline dask/highlevelgraph.py:959]:
```suggestion
        dsk = {}
        deps = {}
        anno = {}
```

**Comment by sjperkins** [comment]
Thanks for working on this @madsbk. I aim to provide feedback later today.

File: dask/highlevelgraph.py
Comment (madsbk) [reply dask/highlevelgraph.py:549]: Not that I have any strong opinion about this but doesn't typeshed specifically state that `Optional` should be used: https://github.com/python/typeshed/blob/master/CONTRIBUTING.md ?

File: dask/highlevelgraph.py
Comment (madsbk) [reply dask/highlevelgraph.py:220]: Fixed

File: dask/highlevelgraph.py
Comment (crusaderky) [reply dask/highlevelgraph.py:550]: My bad. I strongly prefer the implicit version, as IMHO the explicit variant is completely redundant and needlessly verbose.

Comment (madsbk) [reply dask/highlevelgraph.py:550]: I only changed it because `mypy --no-implicit-optional` will complain if `Optional` isn't used. 
But I agree, let's go with the implicit version.

File: dask/highlevelgraph.py
Comment (crusaderky) [inline dask/highlevelgraph.py:219]:
```suggestion
            else:
                if not keys_stringified:
                    keys = [stringify(k) for k in keys]
                    keys_stringified = True
                v = dict.fromkeys(keys, v)

            v.update(annotations.get(a, {}))
            annotations[a] = v
```

Comment (crusaderky) [reply dask/highlevelgraph.py:219]: In case of collision, the original annotations silently win; is it deliberate?

Comment (sjperkins) [reply dask/highlevelgraph.py:219]: I think this may be a case of the original annotations (i.e. supplied as kwargs to Client.{submit, persist} overriding layer annotations. @ian-r-rose may know more here.

File: dask/highlevelgraph.py
Comment (madsbk) [reply dask/highlevelgraph.py:60]: Good point, it should definitely be read-only.

File: dask/highlevelgraph.py
Comment (madsbk) [reply dask/highlevelgraph.py:61]: I am curious, it is recommended to both type hinting the class attribute and `__init__()` as opposed to only type hinting `__init__()` ?
And if yes, should both type hints use `Optional[]`?

File: dask/highlevelgraph.py
Comment (madsbk) [reply dask/highlevelgraph.py:219]: > I think this may be a case of the original annotations (i.e. supplied as kwargs to Client.{submit, persist} overriding layer annotations

Yes, this was also what I was thinking. @ian-r-rose what do you think?

File: dask/highlevelgraph.py
Comment (crusaderky) [reply dask/highlevelgraph.py:61]: yes, they both need to be type hinted as they are not necessarily the same, and mypy makes no assumption as to what to what the init parameters end up used for.
Parameters with a default to None do not need to be wrapped in Optional.

e.g.

```python
class C:
    d: Optional[Dict[str, int]]
    def __init__(self, d: Mapping[str, int] = None):
        self.d = dict(d) if d is not None else None
```

Comment (madsbk) [reply dask/highlevelgraph.py]: :+1:

**Review by sjperkins — COMMENTED** [review]
Thanks for working on this @madsbk. I think this is moving in the right direciton. I have two comments

1. In my mind, annotations are part of the layer state. Is it not possible to :


     * pack annotations into the layer state as suggested in https://github.com/dask/distributed/pull/4406#discussion_r562561455?
   
     *  perform the packing with `__dask_distributed_pack__` and `__dask_distributed_unpack__` rather than separate `__dask_distributed_anno_pack__`?

2. `__dask_distributed_unpack__` returns a tuple which makes future-proofing this function against any custom unpacking difficult. Would it not be possible for it to return a {"dsk": dsk, "deps": deps} dict instead as suggested in https://github.com/dask/distributed/pull/4406#discussion_r562561455?

**Comment by madsbk** [comment]
> 1. In my mind, annotations are part of the layer state. 

I think it is important that we don't force any requirements of the state returned by `__dask_distributed_pack__` other then it must be msg-packable.

Thus, I think we have to decide between two options:

1. `__dask_distributed_pack__` and `__dask_distributed_unpack__` should have the sole responsibility of pack/unpack annotations, which means no automatic filling of an "annotations" field or anything like that. The advantage of this is that we only have one pair of pack/unpack functions. The disadvantage is that all Layer pack implementations will have to call an annotation function or maybe use a function decorator?
2. `__dask_distributed_anno_pack__` and `__dask_distributed_anno_unpack__` should have the sole responsibility of pack/unpack annotations, which makes it possible for Layers to inherent a default implementation. The advantage of this is that most, if not all, Layers don't have to worry about annotations at all. The disadvantage is that we instead of one pair of  pack/unpack functions, have two.

---

> 2. `__dask_distributed_unpack__` returns a tuple which makes future-proofing this function against any custom unpacking difficult.

Good point, returning a dict makes sense.

**Comment by sjperkins** [comment]
> > 1. In my mind, annotations are part of the layer state.
> 
> I think it is important that we don't force any requirements of the state returned by `__dask_distributed_pack__` other then it must be msg-packable.

I think it's reasonable to expect that annotations should be msgpackable -- I would consider it an anti-pattern to use arbitrary literals of complex types i.e. ndarray's etc. in annotations. I think that basic Python types should be good enough for annotaitons -- hence msgpack should be fine.

> 
> Thus, I think we have to decide between two options:
> 
>  1. `__dask_distributed_pack__` and `__dask_distributed_unpack__` should have the sole responsibility of pack/unpack annotations, which means no automatic filling of an "annotations" field or anything like that. The advantage of this is that we only have one pair of pack/unpack functions.
> The disadvantage is that all Layer pack implementations will have to call an annotation function or maybe use a function decorator?

They'd have to call  `__dask_distributed_{pack,unpack}__` on the parent class. to fill annotations. I think this is manageable if  `__dask_distributed_{pack, unpack}__` returns dicts by default as suggested in https://github.com/dask/distributed/pull/4406#discussion_r562561455. There are already calls to parent classes in `__dask_distributed_pack__` in Layer sub-classes IIRC. 
> 
>     2. `__dask_distributed_anno_pack__` and `__dask_distributed_anno_unpack__` should have the sole responsibility of pack/unpack annotations, which makes it possible for Layers to inherent a default implementation. The advantage of this is that most, if not all, Layers don't have to worry about annotations at all. The disadvantage is that we instead of one pair of  pack/unpack functions, have two.

I think this case may be strengthened by an example of custom annotation packing/unpacking for a sub-classed layer, but I can't think of one. Can you think of a counter-example?

If not, I don't think the default annotation packing and unpacking behaviour will change therefore placing the behaviour in `Layer.__dask_distributed_{pack,unpack}__` seems the best approach to me.

**Comment by madsbk** [comment]
> If not, I don't think the default annotation packing and unpacking behaviour will change therefore placing the behaviour in `Layer.__dask_distributed_{pack,unpack}__` seems the best approach to me.

I don't have an example and I don't have any strong opinion against including annotation in `Layer.__dask_distributed_{pack,unpack}__` but notice that in https://github.com/dask/distributed/pull/4406#discussion_r562561455 you still need to call `self.pack_annotations` and `cls.expand_annotations`, which corresponds to `Layer.__dask_distributed_anno_{pack,unpack}__` in this implementation.

I really like the symmetry of `_anno_{pack,unpack}__` but it might be OK to have `Layer.__dask_distributed_{pack,unpack}__` calling them explicitly.

**Comment by sjperkins** [comment]
> > If not, I don't think the default annotation packing and unpacking behaviour will change therefore placing the behaviour in `Layer.__dask_distributed_{pack,unpack}__` seems the best approach to me.
> 
> I don't have an example and I don't have any strong opinion against including annotation in `Layer.__dask_distributed_{pack,unpack}__` but notice that in [dask/distributed#4406 (comment)](https://github.com/dask/distributed/pull/4406#discussion_r562561455) you still need to call `self.pack_annotations` and `cls.expand_annotations`, which corresponds to `Layer.__dask_distributed_anno_{pack,unpack}__` in this implementation.

I don't have strong feelings about the retention of `pack_annotations` and `expand_annotations`. I think moving their functionality into `Layer.__dask_distributed_{pack, unpack}__` would work.

> I really like the symmetry of `_anno_{pack,unpack}__` but it might be OK to have `Layer.__dask_distributed_{pack,unpack}__` calling them explicitly.

I think the symmetry  would be retained by placing their functionality it `Layer.__dask_distributed_{pack,unpack}__`.

**Comment by madsbk** [comment]
> I don't have strong feelings about the retention of `pack_annotations` and `expand_annotations`. I think moving their functionality into `Layer.__dask_distributed_{pack, unpack}__` would work.

I am not sure I follow you, `Layer.__dask_distributed_{pack, unpack}__` are doing a lot of work beside annotations so having Layers call them to handle annotations seams wrong?

Generally, I think it is a bad idea to force `__dask_distributed_{pack, unpack}__` to return dicts in order for the default annotation to work.

I suggest that we either use the current design where `HighLevelGraph.__dask_distributed_unpack__` calls `Layer.__dask_distributed_anno_{pack,unpack}__`
or 
we have all `__dask_distributed_{pack, unpack}__` implementations call `Layer.__dask_distributed_anno_{pack,unpack}__`.

**Comment by sjperkins** [comment]
> > I don't have strong feelings about the retention of `pack_annotations` and `expand_annotations`. I think moving their functionality into `Layer.__dask_distributed_{pack, unpack}__` would work.
> 
> I am not sure I follow you, `Layer.__dask_distributed_{pack, unpack}__` are doing a lot of work beside annotations so having Layers call them to handle annotations seams wrong?

Ah I see where I'm not understanding you: Currently, the default distributed pack/unpack  functionality for the materialized case has been placed in the Layer class. I thought the  `Layer.__dask_distributed_{pack,unpack}__` functions would be unimplemented and the aforementioned functionality placed in BasicLayer.

Would it be possible move the functionality into the BasicLayer? This could be called a MaterializedLayer to make it clearer but I've no strong preference. Then, the base  `Layer.__dask_distributed_{pack,unpack}__` can handle annotation packing/unpacking and subclasses can defer to these methods to achieve this.

**Comment by madsbk** [comment]
> Would it be possible move the functionality into the BasicLayer? This could be called a MaterializedLayer to make it clearer but I've no strong preference. Then, the base `Layer.__dask_distributed_{pack,unpack}__` can handle annotation packing/unpacking and subclasses can defer to these methods to achieve this.

I guess this is possible but I am not sure why? I mean, now that we have everything in Dask doesn't it make sense to have the base class `Layer` implement the default pack/unpack methods? It fits nicely with the Layer class hierarchy already in place.

File: dask/highlevelgraph.py
Comment (ian-r-rose) [reply dask/highlevelgraph.py:219]: > I think this may be a case of the original annotations (i.e. supplied as kwargs to Client.{submit, persist} overriding layer annotations.

Sorry for the slow reply here. Yes, that was my understanding as well. *However*, I don't have a strong intuition of which *should* win, and I can think of an argument for both (more specific annotation vs later-applied annotation).

File: dask/highlevelgraph.py
Comment (ian-r-rose) [reply dask/highlevelgraph.py:550]: Hopefully [PEP-645](https://www.python.org/dev/peps/pep-0645/) will make this a bit more ergonomic.

File: dask/highlevelgraph.py
Comment (ian-r-rose) [reply dask/highlevelgraph.py]: > v is always a dict that you just created

I don't think this is true -- the annotation can be anything msgpack-serializable, and may be just a user-defined value.

**Comment by sjperkins** [comment]
> > Would it be possible move the functionality into the BasicLayer? This could be called a MaterializedLayer to make it clearer but I've no strong preference. Then, the base `Layer.__dask_distributed_{pack,unpack}__` can handle annotation packing/unpacking and subclasses can defer to these methods to achieve this.
> 
> I guess this is possible but I am not sure why? I mean, now that we have everything in Dask doesn't it make sense to have the base class `Layer` implement the default pack/unpack methods? It fits nicely with the Layer class hierarchy already in place.

Let me dive into the code a bit deeper -- I want to base a PR ontop of yours to try out what I'm suggesting and for further discussion.

**Comment by sjperkins** [comment]
@madsbk Here's my suggestion, I wasn't able to use your fork as a base for a PR:

https://github.com/sjperkins/dask/commit/7fcd644738c1652b21f2d6ff363b10b6674a2736

Summary of changes:

1. BasicLayer renamed to MaterialisedLayer.
2. MaterialisedLayer contains code for packing materialized graphs (a.k.a. python dicts) to the scheduler (previously in Layer)
3. `__dask_distributed_{pack, unpack}__` now calls `super()` to pack/unpack annotations.
4. `__dask_distributed_anno_{pack,unpack} removed.

More broadly, MaterialisedLayer is now the single point of responsibility for transmitting materialised graphs. Layer returns to a more abstract form (is_materialised return False).

`py.test -s -vvv distributed/protocol/tests/test_highlevelgraph.py` succeeds for me on the distributed branch.

**Comment by sjperkins** [comment]
> I have a hard time seeing the advantages of your approach. Personally, I was hoping that we could avoid introducing a new protocol/requirements for the packed state returned by `__dask_distributed_pack__()`. Now that we have a class hierarchy, why not handle the pack/unpack through inheritance without having to define a shared state dict?

I think I understand that you're asking why the materialised packing behaviour has been moved into MaterialisedLayer? If so, its an attempt to separate interface from implementation. In that sense, Layer is the interface while the derived classes are implementations of that interface.  The reason for this is that Implementing materialised behaviour in the interface (base class) could potentially constrain the behaviour of derived classes in future: better to place in the behaviour in a specific implementation class.

I get that it may seem like quibbling, but inheritance hierarchies can get [out of hand](https://softwareengineering.stackexchange.com/a/260354).

**Comment by madsbk** [comment]
> I get that it may seem like quibbling, but inheritance hierarchies can get [out of hand](https://softwareengineering.stackexchange.com/a/260354).

I agree inheritance hierarchies can get out of hand but now that we have a class hierarchy and we have a perfectly suited default implementation that is compatible with any conceivable derived class, I think it makes sense. Particularly, if the alternative is an introduction of a key scheme for the packed state dict.

**Comment by sjperkins** [comment]
> > I get that it may seem like quibbling, but inheritance hierarchies can get [out of hand](https://softwareengineering.stackexchange.com/a/260354).
> 
> I agree inheritance hierarchies can get out of hand but now that we have a class hierarchy and we have a perfectly suited default implementation that is compatible with any conceivable derived class, I think it makes sense. Particularly, if the alternative is an introduction of a key scheme for the packed state dict.

OK, I see your point. I'm happy with the PR as it stands then.

**Comment by madsbk** [comment]
@sjperkins, thanks for the discussion I have added your suggestions to let `__dask_distributed_unpack__` return a dict and renamed `BasicLayer -> MaterializedLayer`.

Also @crusaderky, @ian-r-rose, thanks for the reviews.

@jrbourbeau, I think this is ready to be merged after the removal of the CI hack. Could you take a look and if you agree, we should revert `continuous_integration/scripts/install.sh` and merge this PR **and** https://github.com/dask/distributed/pull/4489

**Comment by madsbk** [comment]
@jrbourbeau, I think we should merge this soon. Can you revert `continuous_integration/scripts/install.sh` and merge this PR and dask/distributed#4489 ?

File: dask/highlevelgraph.py
Comment (jrbourbeau) [inline dask/highlevelgraph.py:176]: Similar comment here about the method name:

```suggestion
    def __dask_distributed_annotations_unpack__(
```

File: dask/highlevelgraph.py
Comment (jrbourbeau) [inline dask/highlevelgraph.py:150]: I personally find it much more clear when we use the full `annotations` name instead of using `anno` as a shorthand. My thinking is the name of this method is long either way, so we might as well make it totally clear too

```suggestion
    def __dask_distributed_annotations_pack__(self) -> Optional[Mapping[str, Any]]:
```

File: dask/highlevelgraph.py
Comment (jrbourbeau) [inline dask/highlevelgraph.py:974]:
```suggestion
        return {"dsk": dsk, "deps": deps, "annotations": anno}
```

**Review by jrbourbeau — COMMENTED** [review]
Apologies for the delayed review, thank you for the ping @madsbk. Overall this PR and https://github.com/dask/distributed/pull/4489 look good to me. My primary comment is that there are a few places where I think using the full `annotations` term instead of `anno` for shorthand will improve code clarity. Otherwise, I think we're good to revert the temporary CI changes here and merge.

**Comment by jrbourbeau** [comment]
Hmm actually it looks like there's an issue with how we're handling annotations (see CI over in https://github.com/dask/distributed/pull/4489 -- for example [this CI build](https://github.com/dask/distributed/pull/4489/checks?check_run_id=1975821596))

**Review by jrbourbeau — APPROVED** [review]
Thanks @madsbk! This, and [the companion PR in `distributed`](https://github.com/dask/distributed/pull/4489), look good to me : )

Since we _may_ be releasing tomorrow (xref https://github.com/dask/community/issues/129#issuecomment-786332754), and the changes here are an internal refactor instead of addressing user-facing issues, I'm inclined to wait a day before merging these PRs (until after the release) to give the changes here a little time to simmer in the main branch before pushing them out to users

**Comment by madsbk** [comment]
> Since we _may_ be releasing tomorrow (xref [dask/community#129 (comment)](https://github.com/dask/community/issues/129#issuecomment-786332754)), and the changes here are an internal refactor instead of addressing user-facing issues, I'm inclined to wait a day before merging these PRs (until after the release) to give the changes here a little time to simmer in the main branch before pushing them out to users

Sounds good!

**Comment by jrbourbeau** [comment]
Note I've opened up a follow-up PR at https://github.com/dask/dask/pull/7279 to resolve a CI issue

**Comment by jakirkham** [comment]
Thanks James! 😄

**Pull request merged**
